source/comp/markv_codec.cpp

```diff
@@ -886,141 +927,155 @@ void MarkvCodecBase::ProcessCurInstruction() {
-    const uint32_t size = inst_.words[3];
-    if (multi_mtf_.HasValue(GetMtfIdGeneratedByOpcode(SpvOpTypeFloat),
-                            component_type_id)) {
+    if (opcode == SpvOpTypeFloat) {
+      multi_mtf_.Insert(kMtfTypeScalar, inst_.result_id);
       multi_mtf_.Insert(kMtfTypeFloatScalarOrVector, inst_.result_id);
-    } else if (multi_mtf_.HasValue(GetMtfIdGeneratedByOpcode(SpvOpTypeInt),
-                                   component_type_id)) {
-      multi_mtf_.Insert(kMtfTypeIntScalarOrVector, inst_.result_id);
-    } else if (multi_mtf_.HasValue(GetMtfIdGeneratedByOpcode(SpvOpTypeBool),
-                                   component_type_id)) {
+    }
+
+    if (opcode == SpvOpTypeBool) {
+      multi_mtf_.Insert(kMtfTypeScalar, inst_.result_id);
       multi_mtf_.Insert(kMtfTypeBoolScalarOrVector, inst_.result_id);
     }
-    multi_mtf_.Insert(GetMtfTypeVectorOfSize(size), inst_.result_id);
-  }
 
-  if (inst_.opcode == SpvOpTypeFunction) {
-    const uint32_t return_type = inst_.words[2];
-    multi_mtf_.Insert(kMtfTypeReturnedByFunction, return_type);
-    multi_mtf_.Insert(GetMtfFunctionTypeWithReturnType(return_type),
-                      inst_.result_id);
-  }
+    if (opcode == SpvOpTypeVector) {
+      const uint32_t component_type_id = inst_.words[2];
+      const uint32_t size = inst_.words[3];
+      if (multi_mtf_.HasValue(GetMtfIdGeneratedByOpcode(SpvOpTypeFloat),
+                              component_type_id)) {
+        multi_mtf_.Insert(kMtfTypeFloatScalarOrVector, inst_.result_id);
+      } else if (multi_mtf_.HasValue(GetMtfIdGeneratedByOpcode(SpvOpTypeInt),
+                                     component_type_id)) {
+        multi_mtf_.Insert(kMtfTypeIntScalarOrVector, inst_.result_id);
+      } else if (multi_mtf_.HasValue(GetMtfIdGeneratedByOpcode(SpvOpTypeBool),
+                                     component_type_id)) {
+        multi_mtf_.Insert(kMtfTypeBoolScalarOrVector, inst_.result_id);
+      }
+      multi_mtf_.Insert(GetMtfTypeVectorOfSize(size), inst_.result_id);
+    }
 
-  if (inst_.type_id) {
-    const Instruction* type_inst = FindDef(inst_.type_id);
-    assert(type_inst);
+    if (inst_.opcode == SpvOpTypeFunction) {
+      const uint32_t return_type = inst_.words[2];
+      multi_mtf_.Insert(kMtfTypeReturnedByFunction, return_type);
+      multi_mtf_.Insert(GetMtfFunctionTypeWithReturnType(return_type),
+                        inst_.result_id);
+    }
 
-    multi_mtf_.Insert(kMtfObject, inst_.result_id);
+    if (inst_.type_id) {
+      const Instruction* type_inst = FindDef(inst_.type_id);
+      assert(type_inst);
 
-    multi_mtf_.Insert(GetMtfIdOfType(inst_.type_id), inst_.result_id);
+      multi_mtf_.Insert(kMtfObject, inst_.result_id);
 
-    if (multi_mtf_.HasValue(kMtfTypeFloatScalarOrVector, inst_.type_id)) {
-      multi_mtf_.Insert(kMtfFloatScalarOrVector, inst_.result_id);
-    }
+      multi_mtf_.Insert(GetMtfIdOfType(inst_.type_id), inst_.result_id);
```

Comment:
It's hard to judge the heuristic aspect of the changes.  I'll have to trust you.   But I would be interested in the change in the quality of the result.